### PR TITLE
Speeding up FnCAS ~10x by moving from `nasm` to `as` 

### PR DESCRIPTION
**Pull request by dkorolev**
Had I had more time, I'd plot these. But they seem to be self-explanatory:

```
$ for DIM in $(seq 3 12) ; do for BINARY in optimize_as optimize_clang optimize_nasm ; do echo $BINARY"@"$DIM ; ./$BINARY --dim $DIM --logtostderr 2>&1 | grep seconds ; echo ; done ; done | tee report.txt

optimize_as@3
Optimizer: Done compiling the objective function, took 0.064243 seconds.
Optimizer: Done compiling the gradient, took 0.179195 seconds.

optimize_clang@3
Optimizer: Done compiling the objective function, took 0.060117 seconds.
Optimizer: Done compiling the gradient, took 0.528718 seconds.

optimize_nasm@3
Optimizer: Done compiling the objective function, took 0.021945 seconds.
Optimizer: Done compiling the gradient, took 0.794939 seconds.

optimize_as@4
Optimizer: Done compiling the objective function, took 0.065677 seconds.
Optimizer: Done compiling the gradient, took 0.458750 seconds.

optimize_clang@4
Optimizer: Done compiling the objective function, took 0.064909 seconds.
Optimizer: Done compiling the gradient, took 1.529436 seconds.

optimize_nasm@4
Optimizer: Done compiling the objective function, took 0.027482 seconds.
Optimizer: Done compiling the gradient, took 2.559039 seconds.

optimize_as@5
Optimizer: Done compiling the objective function, took 0.063807 seconds.
Optimizer: Done compiling the gradient, took 0.944710 seconds.

optimize_clang@5
Optimizer: Done compiling the objective function, took 0.072828 seconds.
Optimizer: Done compiling the gradient, took 3.887104 seconds.

optimize_nasm@5
Optimizer: Done compiling the objective function, took 0.036113 seconds.
Optimizer: Done compiling the gradient, took 6.579334 seconds.

optimize_as@6
Optimizer: Done compiling the objective function, took 0.130903 seconds.
Optimizer: Done compiling the gradient, took 1.867289 seconds.

optimize_clang@6
Optimizer: Done compiling the objective function, took 0.067057 seconds.
Optimizer: Done compiling the gradient, took 8.850630 seconds.

optimize_nasm@6
Optimizer: Done compiling the objective function, took 0.046722 seconds.
Optimizer: Done compiling the gradient, took 14.606663 seconds.

optimize_as@7
Optimizer: Done compiling the objective function, took 0.070455 seconds.
Optimizer: Done compiling the gradient, took 3.501311 seconds.

optimize_clang@7
Optimizer: Done compiling the objective function, took 0.113944 seconds.
Optimizer: Done compiling the gradient, took 17.354487 seconds.

optimize_nasm@7
Optimizer: Done compiling the objective function, took 0.054269 seconds.
Optimizer: Done compiling the gradient, took 28.425242 seconds.

optimize_as@8
Optimizer: Done compiling the objective function, took 0.047216 seconds.
Optimizer: Done compiling the gradient, took 6.272984 seconds.

optimize_clang@8
Optimizer: Done compiling the objective function, took 0.076810 seconds.
Optimizer: Done compiling the gradient, took 31.519345 seconds.

optimize_nasm@8
Optimizer: Done compiling the objective function, took 0.031777 seconds.
Optimizer: Done compiling the gradient, took 53.151395 seconds.

optimize_as@9
Optimizer: Done compiling the objective function, took 0.067737 seconds.
Optimizer: Done compiling the gradient, took 10.764020 seconds.

optimize_clang@9
Optimizer: Done compiling the objective function, took 0.085724 seconds.
Optimizer: Done compiling the gradient, took 53.565371 seconds.

optimize_nasm@9
Optimizer: Done compiling the objective function, took 0.075348 seconds.
Optimizer: Done compiling the gradient, took 92.410523 seconds.

optimize_as@10
Optimizer: Done compiling the objective function, took 0.073644 seconds.
Optimizer: Done compiling the gradient, took 17.695206 seconds.

optimize_clang@10
Optimizer: Done compiling the objective function, took 0.092457 seconds.
Optimizer: Done compiling the gradient, took 87.924109 seconds.

optimize_nasm@10
Optimizer: Done compiling the objective function, took 0.092804 seconds.
Optimizer: Done compiling the gradient, took 152.048166 seconds.

optimize_as@11
Optimizer: Done compiling the objective function, took 0.079173 seconds.
Optimizer: Done compiling the gradient, took 27.668103 seconds.

optimize_clang@11
Optimizer: Done compiling the objective function, took 0.095975 seconds.
Optimizer: Done compiling the gradient, took 161.110257 seconds.

optimize_nasm@11
Optimizer: Done compiling the objective function, took 0.145706 seconds.
Optimizer: Done compiling the gradient, took 236.959049 seconds.

optimize_as@12
Optimizer: Done compiling the objective function, took 0.180046 seconds.
Optimizer: Done compiling the gradient, took 43.333109 seconds.

optimize_clang@12
Optimizer: Done compiling the objective function, took 1.720292 seconds.

optimize_nasm@12
Optimizer: Done compiling the objective function, took 0.250477 seconds.
Optimizer: Done compiling the gradient, took 364.147382 seconds.
```

Thanks,
Dima

### CODE REVIEW

**Comment by mzhurovich** [comment]
@dkorolev It doesn't compile on Mac, but somehow it results in `ALL TESTS PASS` :-/
https://travis-ci.org/C5T/Current/jobs/179339548
```
[ RUN      ] FnCAS.TrivialCompiledFunctions
/tmp/.current-tmp-000041a7.s:1:15: error: unexpected token in '.section' directive
.section .text
              ^
gcc -O0 -shared /tmp/.current-tmp-000041a7.s -o /tmp/.current-tmp-000041a7.so -lm
256
make[1]: *** [test] Error 255
...
ALL TESTS PASS.

The command "make individual_tests" exited with 0.
```

**Comment by mzhurovich** [comment]
@dkorolev Here is the quick and dirty fix for OS X: https://github.com/C5T/Current/commit/6796d72431f99b5e59a26704435c98c7fd3a379c
Introducing the platform dependent name prefixing routines would be a cleaner solution.

**Comment by dkorolev** [comment]
@mzhurovich -- hope this PR is good to go in its present state.

Another small yet important thought I'm pondering on is a template parameter to set optimization direction. But it's certainly not going to happen today.

Thanks!
Dima

**Comment by dkorolev** [comment]
PS1: The follow-up to the `"Fixed" `individual-tests.sh` for Travis to actually fail; should be red` commit did the job, I tested. It must be force-push that confused GitHub to believe it's green.

PS2: Thanks for your commit with the Mac OS fix!

Dima

**Comment by mzhurovich** [comment]
Speaking of external early stopping criteria - why not pass `PointAndValue` + gradient?

**Comment by dkorolev** [comment]
:+1: 274a9d1

File: examples/Optimize/optimize.cc [hunk 1]
```diff
@@ -42,6 +42,7 @@ DEFINE_int32(min, 1, "The minimum randomly generated value for A[i][j].");
 DEFINE_int32(max, 9, "The minimum randomly generated value for A[i][j].");
 DEFINE_uint64(seed, 19830816, "The random seed to use, zero to generate the input nondeterministically.");
 DEFINE_double(eps, 1e-3, "The value of the epsilon to compare doubles with.");
+DEFINE_bool(nojit, false, "Set to disable JIT and run the optimization using FnCAS' intermediate expresssion format.");
```
Comment (mzhurovich) [inline examples/Optimize/optimize.cc]: `s/sss/ss/`

Comment (dkorolev) [inline examples/Optimize/optimize.cc]: Никогда ещё Штирлиц не был так близок к провалу.

**Comment by dkorolev** [comment]
And I have no idea why does the ["Print JIT name and test CLANG."](https://github.com/C5T/Current/pull/695/commits/43b54179557b4dc281c5bd10e7f3ca57f15eeb0c) commit  in this GitHub view of the pull request point to an unrelated [Travis](https://travis-ci.org/C5T/Current/builds/179996127) build that shows `FNCAS_JIT` as `AS`, not `CLANG`.

I've pushed a new branch to re-run the `CLANG` test on Travis ([branch](https://github.com/dkorolev/Current/tree/fncas_test_clang_on_travis), [Travis](https://travis-ci.org/dkorolev/Current/builds/180126400)). Suggest we also test `AS` vs. `CLANG` on Linux and Mac manually.

**Comment by mzhurovich** [comment]
@dkorolev Yo!

**Pull request merged**